### DISCUSSION TITLE
When will PDF signing be supported?

### DISCUSSION THREAD

**Discussion by JorjMcKie** [pymupdf/PyMuPDF#757] [Q&A]
Signing PDF document signing is an important feature - and the base library MuPDF supports it.
However, PyMuPDF **_does not yet_**.

The reasons:

* We know that MuPDF must be installed before PyMuPDF can be built. These two steps both happen when PyMuPDF's wheels are being created (using the great Travis CI service).
* MuPDF installation material is complete and self-sufficient: it contains all its required thirdparty software (like ZIP, Freetype, OpenJPEG, Harfbuzz, etc.) in folder ``thirdparty``.
* The previous is **_not true for OpenSSL_**: it must be installed separately **_before_** installing MuPDF. OpenSSL is a prerequisite for MuPDF to support PDF signing. Unfortunately, there does not seem to exist a common installation procedure for all three wheel platforms OSX, Linux and Windows.
* A well-defined, **_repeatable_** installation approach is however required, before I can even think of supporting PDF document signing in PyMuPDF.

**Comment by hbh112233abc** [comment]
Dear JoriMcKie
I think you can remark how to sign pdf with pymupdf in the guide document.
And the guide document show use sign function must install OpenSSL before

**Comment by jloehel** [comment]
How can I support you?

**Reply by JorjMcKie** [reply]
Most valuable would be an installtion procedure from sources on platform Linux:
* download & install OpenSSL
* download MuPDF
* determine MuPDF installation parameter adjustments for OpenSSL
* install MuPDF
* download & setp.py PyMuPDF sources

Once such a script works it would be no major problem to make a GitHub Action from it, that generates wheels.
It should also be no rocket science to port such a script to Mac OSX.

As far as Windows is concerned, I would take of it.

**Reply by jloehel** [reply]
> Most valuable would be an installtion procedure from sources on platform Linux:
> 
>     * download & install OpenSSL
> 
>     * download MuPDF
> 
>     * determine MuPDF installation parameter adjustments for OpenSSL
> 
>     * install MuPDF
> 
>     * download & setp.py PyMuPDF sources
> 
> 
> Once such a script works it would be no major problem to make a GitHub Action from it, that generates wheels. It should also be no rocket science to port such a script to Mac OSX.
> 
> As far as Windows is concerned, I would take of it.

What Linux do you use for building? Ubuntu? In the spec for OpenSUSE it's enough to add openssl to the build requirements: https://build.opensuse.org/package/view_file/Publishing/mupdf/mupdf.spec?expand=1

**Reply by JorjMcKie** [reply]
On my machine I use WSL Ubuntu. Github Actions the same.

**Comment by jloehel** [comment]
Hi @JorjMcKie,

Sorry for the really long delay. I looked into the signing thingie but I realized that PDF document signing with mupdf is a little bit hacky. The document needs already the necessary objects sig, stamp objects in-place. I am using now pyHanko for signing the documents. It's much easier to create a stamp for the documents after the creation with pymupdf.
